pkg/prober/prober.go

```diff
@@ -0,0 +1,152 @@
+	"net/url"
+	"os"
+	"syscall"
+	"time"
+)
+
+type Prober interface {
+	Probe(context.Context, url.URL) (time.Duration, error)
+	String() string
+}
+
+// The Client interface's purpose is to enable tests of the Probers that consume a Client
+// instead of a http.Client.
+type Client interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// NoProber implements the Probeer interface.
+// NoProber will always return an error and a large duration.
+type NoProber struct{}
+
+func (p *NoProber) Probe(ctx context.Context, u url.URL) (time.Duration, error) {
+	dur := time.Duration(1<<63 - 1)
+	return dur, errors.New("this is no probe")
+}
+
+func (p *NoProber) String() string {
+	return "no-prober"
+}
+
+// HttpPingProber implements the Prober interface.
+type HttpPingProber struct {
+	c Client
+}
+
+// NewHttpPingProber returns a HttpProber.
+// It is safe to use concurrently.
+func NewHttpPingProber(c Client) *HttpPingProber {
+	return &HttpPingProber{c: c}
+}
+
+func (p *HttpPingProber) Probe(ctx context.Context, u url.URL) (time.Duration, error) {
+	u.Path = "ping"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create request: %w", err)
+	}
+	start := time.Now()
+	var dur time.Duration
+	var res *http.Response
+	if res, err = p.c.Do(req); err != nil {
+		return 0, fmt.Errorf("failed to make request to %s: %w", u.String(), err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("expected status Code 200, got %d", res.StatusCode)
+	}
+	dur = time.Now().Sub(start)
+	if _, err := io.Copy(ioutil.Discard, res.Body); err != nil {
+		log.Printf("failed to discard body: %v\n", err)
+	}
+	if err := res.Body.Close(); err != nil {
+		log.Printf("failed to close body: %v\n", err)
+	}
+	return dur, nil
+}
+
+func (p *HttpPingProber) String() string {
+	return "http-ping-prober"
+}
+
+// HttpProber implements the Prober interface.
+type HttpProber struct {
+	c Client
+}
+
+// NewHttpProber returns a HttpProber.
+// It is safe to use concurrently.
+func NewHttpProber(c Client) *HttpProber {
+	return &HttpProber{c: c}
+}
+
+func (p *HttpProber) Probe(ctx context.Context, u url.URL) (time.Duration, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create request: %w", err)
+	}
+	start := time.Now()
+	var dur time.Duration
+	var res *http.Response
+	if res, err = p.c.Do(req); err != nil {
+		return 0, fmt.Errorf("failed to make request to %s: %w", u.String(), err)
+	}
+	dur = time.Now().Sub(start)
+	if _, err := io.Copy(ioutil.Discard, res.Body); err != nil {
+		log.Printf("failed to discard body: %v\n", err)
+	}
+	if err := res.Body.Close(); err != nil {
+		log.Printf("failed to close body: %v\n", err)
+	}
+	return dur, nil
+}
+
+func (p *HttpProber) String() string {
+	return "http-prober"
+}
+
+// TcpProber implements the Prober interface.
+type TcpProber struct {
```

Comment:
```suggestion
type TCPProber struct {
```

Nit: by convention we keep all of the letters of an abbreviation in the same casing in Go. That means that if it's exported, it would be `TCP...` if not exported, it's `tcp...`. For example, your func `NewTCPProber` follows the convention correctly